### PR TITLE
[Clang][HIP] Deprecate the AMDGCN_WAVEFRONT_SIZE macros

### PR DESCRIPTION
**Pull request by ritter-x2a**
So far, these macros can be used in contexts where no meaningful wavefront size is available. We therefore deprecate these macros, to replace them with a more resilient interface to access wavefront size information where it is available.

For SWDEV-491529.

### CODE REVIEW

File: clang/include/clang/Basic/MacroBuilder.h
Comment (arsenm) [inline clang/include/clang/Basic/MacroBuilder.h:33]: Just use Twine with a default ""? Shouldn't have a value twine optional

File: clang/lib/Basic/Targets/AMDGPU.cpp
Comment (arsenm) [inline clang/lib/Basic/Targets/AMDGPU.cpp:345]: I think it is unhelpful to deprecate without a suggested replacement.

I also disagree with removing this for the device side compile. There still are uses for it, and there are still non-offloading languages

Comment (AlexVlx) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: If you are removing an antipattern, the replacement is “don't do this”? Not all things that exists are valid. We should absolutely remove this, at least for HIP, where it creates needless divergence from CUDA and is a bug farm as a consequence.

Orthogonally, what are the uses for it that you have in mind, and which non-offloading languages are you thinking of? The only scenario I can envision here is if one does an end-to-end compile (so not the weird 1-per-target thing we do with offload, and not playing games around single AST that gets forked at CodeGen time). In that case, yes, that *might* be useful, but delivering it as a macro is fundamentally a bad design IMHO, so even there we should not have this. Furthermore, at the moment, that sort of use is incredibly underspecified / governed by happens to work by accident rather than by design, so building some early technical debt seems problematic, at least to me

File: clang/include/clang/Basic/MacroBuilder.h
Comment (ritter-x2a) [reply clang/include/clang/Basic/MacroBuilder.h]: Changed in a798615ab4608f1681b15a526bc898df3c12bfa3.

File: clang/lib/Basic/Targets/AMDGPU.cpp
Comment (jhuber6) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: This never should've been defined on the host side for sure. We do this in the OpenMP device runtime.
```c
return llvm::omp::getAMDGPUGridValues<__AMDGCN_WAVEFRONT_SIZE>();
```
That could probably be replaced with
```c
return __builtin_amdgcn_wavefrontsize();
```

**Review by shiltian — COMMENTED** [review]
Just a heads up, this change is gonna break some OpenMP tests, specifically those prefix with `ompx`.

**Comment by AlexVlx** [comment]
> Just a heads up, this change is gonna break some OpenMP tests, specifically those prefix with `ompx`.

Just to clarify, adding the deprecation warning will break them, or the eventual, as-of-yet not-scheduled, removal, will?

**Comment by ritter-x2a** [comment]
> > Just a heads up, this change is gonna break some OpenMP tests, specifically those prefix with `ompx`.
> 
> Just to clarify, adding the deprecation warning will break them, or the eventual, as-of-yet not-scheduled, removal, will?

On my system, the deprecation warning did not break any lit tests (nor tests from our internal testing). The `ompx` tests use the macro and will break once we actually remove it.

**Comment by jhuber6** [comment]
I made https://github.com/llvm/llvm-project/pull/113156 to hopefully address the OpenMP issue. We use it in the runtime itself for a lot of stuff so removing it would definitely break that. The function call here matches what NVIDIA provides via PTX. I think we should have an optimization that replaces the intrinsic with a constant if the function it belongs to has `wavefrontsize32` or whatever attributes.

**Comment by AlexVlx** [comment]
> I made #113156 to hopefully address the OpenMP issue. We use it in the runtime itself for a lot of stuff so removing it would definitely break that. The function call here matches what NVIDIA provides via PTX. I think we should have an optimization that replaces the intrinsic with a constant if the function it belongs to has `wavefrontsize32` or whatever attributes.

We already do this during legalisation / legacy ISEL; there might be merit into doing it as early as possible though.

**Comment by shiltian** [comment]
> > > Just a heads up, this change is gonna break some OpenMP tests, specifically those prefix with `ompx`.
> > 
> > 
> > Just to clarify, adding the deprecation warning will break them, or the eventual, as-of-yet not-scheduled, removal, will?
> 
> On my system, the deprecation warning did not break any lit tests (nor tests from our internal testing). The `ompx` tests use the macro and will break once we actually remove it.

Oh didn’t notice this PR was actually to just warn it instead of entirely removing it. Sorry for the notice.

File: clang/lib/Basic/Targets/AMDGPU.cpp
Comment (arsenm) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: I was thinking of lower level uses. i.e. not user facing code we would want in headers or people using directly. e.g. cases where implementing cross lane operations relies on the operations that only work on half the lanes, or don't work at all when executed in wave32

**Comment by arsenm** [comment]
> We already do this during legalisation / legacy ISEL; there might be merit into doing it as early as possible though.

Yes, we definitely need to start folding this out much earlier.

**Comment by ritter-x2a** [comment]
Should we move on with the deprecation of the macros with this PR then? Please let me know if there are technical concerns remaining with the PR or approve it so that it can land in trunk.

**Comment by jhuber6** [comment]
> Should we move on with the deprecation of the macros with this PR then? Please let me know if there are technical concerns remaining with the PR or approve it so that it can land in trunk.

CI is red but it seems to be some weird Windows failures. Ideally if the builtin / intrinsic is the way to go I'd like that to be folded much earlier because it prevents loop unrolling for cases like SIMT scan / reduce.

**Comment by AlexVlx** [comment]
> > Should we move on with the deprecation of the macros with this PR then? Please let me know if there are technical concerns remaining with the PR or approve it so that it can land in trunk.
> 
> CI is red but it seems to be some weird Windows failures. Ideally if the builtin / intrinsic is the way to go I'd like that to be folded much earlier because it prevents loop unrolling for cases like SIMT scan / reduce.

I should have a patch for earlier folding up today or tomorrow, so that will hopefully be addressed.

**Comment by ritter-x2a** [comment]
> > Should we move on with the deprecation of the macros with this PR then? Please let me know if there are technical concerns remaining with the PR or approve it so that it can land in trunk.
> 
> CI is red but it seems to be some weird Windows failures. Ideally if the builtin / intrinsic is the way to go I'd like that to be folded much earlier because it prevents loop unrolling for cases like SIMT scan / reduce.

Let me rebase so that CI runs again; Windows CI was a bit of a mess in the last week or so.

**Comment by ritter-x2a** [comment]
> > > Should we move on with the deprecation of the macros with this PR then? Please let me know if there are technical concerns remaining with the PR or approve it so that it can land in trunk.
> > 
> > 
> > CI is red but it seems to be some weird Windows failures. Ideally if the builtin / intrinsic is the way to go I'd like that to be folded much earlier because it prevents loop unrolling for cases like SIMT scan / reduce.
> 
> Let me rebase so that CI runs again; Windows CI was a bit of a mess in the last week or so.

CI looks better now.

**Comment by AlexVlx** [comment]
> > CI is red but it seems to be some weird Windows failures. Ideally if the builtin / intrinsic is the way to go I'd like that to be folded much earlier because it prevents loop unrolling for cases like SIMT scan / reduce.

#114481 should deal with this @arsenm @jhuber6.

**Comment by ritter-x2a** [comment]
@AlexVlx @jhuber6 @arsenm  is there a dependence between this deprecation PR and #114481, or can we already go ahead with the deprecation of the macro via this PR?

**Comment by jhuber6** [comment]
This just emits a warning so it doesn't actually change anything yet. I'm waiting on the wavesize folding before I land the changes in `libomptarget` because it'd have performance considerations.

**Review by AlexVlx — APPROVED** [review]
LGTM, thanks.

File: clang/test/Driver/hip-wavefront-size-deprecation-diagnostics.hip
Comment (nico) [inline clang/test/Driver/hip-wavefront-size-deprecation-diagnostics.hip:7]: 1. Clang's tests must be hermetic and can't include system headers
2. This fails on Mac: http://45.33.8.238/macm1/95266/step_6.txt

Please take a look and revert for now if it takes a while to fix.

Comment (ritter-x2a) [reply clang/test/Driver/hip-wavefront-size-deprecation-diagnostics.hip:7]: @nico Thanks for the report and sorry for the inconvenience! I reverted the patch with #115499 and opened PR #115507 with a fix. I'd appreciate it if you could take a look and see if that fixes the issue.

Comment (nico) [reply clang/test/Driver/hip-wavefront-size-deprecation-diagnostics.hip:7]: Thanks! Since the test is now hermetic, I'm guessing it'll work now. I'll check my bot after you've relanded your patch.

File: clang/lib/Basic/Targets/AMDGPU.cpp
Comment (al42and) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: The use we currently have in our project is to detect, via CMake's `try_compile`, whether we have Wave64 or Wave32 targets (or both) to later choose which data structure flavors to instantiate. With the macro, we can easily do `#if`+`#warning` inside a kernel, and then parse the build log.

Without it, the only solution we found was to parse the list of targets, which feels more brittle. `__builtin_amdgcn_wavefrontsize()` is not a compile-time constant expression, so cannot be used for such checks. Am I missing something? What is the planned timeline for the removal of this macro?

Comment (jhuber6) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: The problem is that it's not that simple, since the user can easily change the wavefront size by compiling with `-mwavefrontsize64`. I ran into these types of issues myself while working on the RPC interface for the [rpc](https://github.com/llvm/llvm-project/blob/main/libc/shared/rpc.h#L339) interface. There I pretty much just take the wavefront size as an argument to the indexing / allocation functions. Alternatively you use a template and do runtime dispatch and assume that optimizations will DCE the other code.

Comment (jhuber6) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: Personally, I think these macros are fine for the case where the user manually passed an `-mcpu=` argument to the compilation job, and we're not targeting SPIR-V. As far as I know, the main motivation is because for HIP SPIR-V it's an incorrect value since it will JIT compile, but if we're not doing that then it's fine.

Comment (AlexVlx) [reply clang/lib/Basic/Targets/AMDGPU.cpp:345]: This is incorrect, the change was not motivated by anything SPIR-V related, the discussion predates it.

**Pull request merged**
